server/config/config.go

```diff
@@ -409,42 +435,75 @@ func (f *structSliceFlag) Set(value string) error {
 	return nil
 }
 
-func defineFlagsForMembers(parentStructNames []string, T reflect.Value) {
+// If configSet is nil, defines all global flags present in the config that have
+// not been elsewhere defined. If configSet isn't nil, defines flags in the
+// provided FlagSet for the config.
+func defineFlagsForMembers(parentStructNames []string, T reflect.Value, configSet *flag.FlagSet) {
 	typeOfT := T.Type()
 	for i := 0; i < T.NumField(); i++ {
 		f := T.Field(i)
 		fieldName := typeOfT.Field(i).Tag.Get("yaml")
-		docString := typeOfT.Field(i).Tag.Get("usage")
 		fqFieldName := strings.ToLower(strings.Join(append(parentStructNames, fieldName), "."))
+		docString := typeOfT.Field(i).Tag.Get("usage")
 
 		switch f.Type().Kind() {
 		case reflect.Ptr:
 			log.Fatal("The config should not contain pointers!")
 		case reflect.Struct:
-			defineFlagsForMembers(append(parentStructNames, fieldName), f)
+			defineFlagsForMembers(append(parentStructNames, fieldName), f, configSet)
 			continue
 		case reflect.Bool:
-			flag.BoolVar(f.Addr().Interface().(*bool), fqFieldName, f.Bool(), docString)
+			if configSet != nil {
+				configSet.BoolVar(f.Addr().Interface().(*bool), fqFieldName, f.Bool(), docString)
+			} else if flag.Lookup(fqFieldName) == nil {
+				flag.Bool(fqFieldName, false, docString)
+			}
 		case reflect.String:
-			flag.StringVar(f.Addr().Interface().(*string), fqFieldName, f.String(), docString)
+			if configSet != nil {
+				configSet.StringVar(f.Addr().Interface().(*string), fqFieldName, f.String(), docString)
+			} else if flag.Lookup(fqFieldName) == nil {
+				flag.String(fqFieldName, "", docString)
+			}
 		case reflect.Int:
-			flag.IntVar(f.Addr().Interface().(*int), fqFieldName, int(f.Int()), docString)
+			if configSet != nil {
+				configSet.IntVar(f.Addr().Interface().(*int), fqFieldName, int(f.Int()), docString)
+			} else if flag.Lookup(fqFieldName) == nil {
+				flag.Int(fqFieldName, 0, docString)
+			}
```

Comment:
flag.CommandLine exposes the "default" flag set (which flag.Int defines, for example).

I think you could significantly simplify this code by removing all the else if cases by assuming a flag set is *always* provided. In the case where you want to set the default flags, you can pass in flag.CommandLine (intead of nil)